core/src/main/java/com/linecorp/armeria/client/Bootstraps.java

```diff
@@ -0,0 +1,136 @@
+                  SerializationFormat serializationFormat) {
+        if (!httpAndHttpsValues().contains(desiredProtocol)) {
+            throw new NullPointerException("Unsupported session protocol: " + desiredProtocol);
+        }
+
+        if (remoteAddress instanceof InetSocketAddress) {
+            return select(inetBootstraps, desiredProtocol, serializationFormat);
+        }
+
+        assert remoteAddress instanceof DomainSocketAddress : remoteAddress;
+
+        if (unixBootstraps == null) {
+            throw new IllegalArgumentException("Domain sockets are not supported by " +
+                                               eventLoop.getClass().getName());
+        }
+
+        return select(unixBootstraps, desiredProtocol, serializationFormat);
+    }
+
+    int getConnectTimeoutMillis() {
+        return connectTimeoutMillis;
+    }
+
+    private Bootstrap select(Bootstrap[][] bootstraps, SessionProtocol desiredProtocol, SerializationFormat serializationFormat) {
+        return bootstraps[desiredProtocol.ordinal()][toIndex(serializationFormat)];
+    }
+
+    private Bootstrap[][] newBootstrapMap(Bootstrap baseBootstrap,
+                                          HttpClientFactory clientFactory,
+                                          EventLoop eventLoop) {
+        baseBootstrap.group(eventLoop);
+        final Set<SessionProtocol> sessionProtocols = httpAndHttpsValues();
+        final Bootstrap[][] maps = (Bootstrap[][]) Array.newInstance(
+                Bootstrap.class, SessionProtocol.values().length, 2);
+        // Attempting to access the array with an unallowed protocol will trigger NPE,
+        // which will help us find a bug.
+        for (SessionProtocol p : sessionProtocols) {
+            final SslContext sslCtx = determineSslContext(p);
+            setBootstrap(baseBootstrap.clone(), clientFactory, maps, p, sslCtx, true);
+            setBootstrap(baseBootstrap.clone(), clientFactory, maps, p, sslCtx, false);
+        }
+        return maps;
+    }
+    private void setBootstrap(Bootstrap bootstrap, HttpClientFactory clientFactory, Bootstrap[][] maps,
+                              SessionProtocol p, SslContext sslCtx, boolean webSocket) {
+        bootstrap.handler(new ChannelInitializer<Channel>() {
+            @Override
+            protected void initChannel(Channel ch) throws Exception {
+                ch.pipeline().addLast(
+                        new HttpClientPipelineConfigurator(clientFactory, webSocket, p, sslCtx));
+            }
+        });
+        maps[p.ordinal()][toIndex(webSocket)] = bootstrap;
+    }
+
+    private int toIndex(boolean webSocket) {
+        return webSocket ? 1 : 0;
+    }
+
+    private int toIndex(SerializationFormat serializationFormat) {
+        return toIndex(serializationFormat == SerializationFormat.WS);
+    }
+
+    private SslContext determineSslContext(SessionProtocol desiredProtocol) {
```

Comment:
Do you mind deduping this with `HttpChannelPool#determineSslContext`? I think `HttpChannelPool` can call `Bootstraps#determineSslContext`

Reply:
That's a good idea. I think using `Bootstraps#determineSslContext` in `HttpConnectionPool` would reduce the fields needed for that method, and also reduce the dedeuping code 😍. Thanks.